builtin/v13/market/invariants.go

```diff
@@ -254,6 +262,65 @@ func CheckStateInvariants(st *State, store adt.Store, balance abi.TokenAmount, c
 
 	acc.Require(len(expectedDealOps) == 0, "missing deal ops for proposals: %v", expectedDealOps)
 
+	//
+	// Provider Sectors
+	// Provider->sector->deal mapping
+	// Each entry corresponds to non-terminated deal state.
+	// A deal may have expired but remain in the mapping until settlement.
+
+	providerSectors := make(map[abi.SectorID][]abi.DealID)
+	if sectorDeals, err := adt.AsMap(store, st.ProviderSectors, ProviderSectorsHamtBitwidth); err != nil {
+		acc.Addf("error loading st.ProviderSectors: %v", err)
+	} else {
+		var sectorMapRoot cbg.CborCid
+		err = sectorDeals.ForEach(&sectorMapRoot, func(providerID string) error {
+			provider, err := abi.ParseUIntKey(providerID)
+			acc.RequireNoError(err, "error getting address from bytes")
+
+			sectorMap, err := adt.AsMap(store, cid.Cid(sectorMapRoot), ProviderSectorsHamtBitwidth)
+			acc.RequireNoError(err, "error loading sector map for provider %s", provider)
+
+			var dealIDs SectorDealIDs
+			err = sectorMap.ForEach(&dealIDs, func(sectorID string) error {
+				sectorNumber, err := abi.ParseUIntKey(sectorID)
+				acc.RequireNoError(err, "error sector number from bytes")
+
+				acc.Require(len(dealIDs) > 0, "no deal ids in sector %v", sectorID)
+
+				dealIDsCopy := make([]abi.DealID, len(dealIDs))
+				copy(dealIDsCopy, dealIDs)
+
+				providerSectors[abi.SectorID{Miner: abi.ActorID(provider), Number: abi.SectorNumber(sectorNumber)}] = dealIDsCopy
+
+				// check against proposalStats
+				for _, dealID := range dealIDsCopy {
+					st, found := proposalStats[dealID]
+					acc.Require(found, "deal id %d in provider sectors not found in proposals", dealID)
+					acc.Require(st.SectorNumber == abi.SectorNumber(sectorNumber), "deal id %d sector number %d does not match sector id %d", dealID, st.SectorNumber, sectorNumber)
+
+					delete(expectedProviderSectors, dealID)
```

Comment:
Validate that it's in this map before deleting it?

Reply:
Yes, thanks!